### ISSUE TITLE
The "km controller-manager" command doesn't work successfully behind proxy.

### ISSUE THREAD

**Issue by NanXiao** [mesosphere/kubernetes-mesos#718]
My testbed is like this: There is one `RHEL 7.0` host, and 3 `Ubuntu` VMs runs on this host.`k8s`, `Mesos Master` and `Mesos Client` runs separately on these VMs. 

`K8s` IP is `10.100.3.41`, and `Mesos Master` is `10.100.3.49`. And I also need configure `proxy`:  

	http_proxy="http://web-proxy.corp.xx.com:8080/"
	https_proxy="https://web-proxy.corp.xx.com:8080/"
	ftp_proxy="http://web-proxy.corp.xx.com:8080/"
	no_proxy="localhost,127.0.0.1,localaddress,.localdomain.com"
	HTTP_PROXY="http://web-proxy.corp.xx.com:8080/"
	HTTPS_PROXY="https://web-proxy.corp.xx.com:8080/"
	FTP_PROXY="http://web-proxy.corp.xx.com:8080/"
	NO_PROXY="localhost,127.0.0.1,localaddress,.localdomain.com" 
 
But I find if I don't add `Mesos Master` IP (`10.100.3.49`) into `no_proxy`, the `km controller-manager` command doesn't run successfully. And there are following errors in controller.log:  
	
	......
	Line 600: E1223 09:32:04.727250    5374 statusupdater.go:68] Error listing slaves without kubelet: Get http://10.100.3.49:5050/state.json: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
	Line 862: E1223 09:32:14.738837    5374 statusupdater.go:68] Error listing slaves without kubelet: Get http://10.100.3.49:5050/state.json: http: error connecting to proxy http://web-proxy.corp.xx.com:8080/: dial tcp: lookup web-proxy.corp.hp.com on 16.155.255.10:53: read udp 10.100.3.41:36156->16.155.255.10:53: i/o timeout
	......
	Line 7928: E1223 09:36:33.037835    5374 statusupdater.go:68] Error listing slaves without kubelet: HTTP request failed with code 403: 403 Forbidden
	Line 8216: E1223 09:36:43.455506    5374 statusupdater.go:68] Error listing slaves without kubelet: HTTP request failed with code 403: 403 Forbidden

And there is only one `TCP` connection with `Mesos Master`:  

	$ sudo lsof -p 6474 -P -n | grep 5050
	km      6474  nan    6u  IPv4  22159      0t0    TCP 10.100.3.41:36886->10.100.3.49:5050 (ESTABLISHED)
	$

But if I add `Mesos Master` IP (`10.100.3.49`) into `no_proxy`, the `km controller-manager` run successfully. And there are two `TCP` connection with `Mesos Master`:  

	$ sudo lsof -p 6532 -P -n | grep 5050
	km      6532  nan    6u  IPv4  22701      0t0    TCP 10.100.3.41:36923->10.100.3.49:5050 (ESTABLISHED)
	km      6532  nan   16u  IPv4  22753      0t0    TCP 10.100.3.41:36942->10.100.3.49:5050 (ESTABLISHED)
	$

In the first case, if I don't add add `Mesos Master` IP (`10.100.3.49`) into `no_proxy`, why there is still one successful `TCP` connection?

**Comment by sttts** [comment]
Is it a typo with xx.com and hp.com: `error connecting to proxy http://web-proxy.corp.xx.com:8080/: dial tcp: lookup web-proxy.corp.hp.com` ?

**Comment by NanXiao** [comment]
@sttts Yes! I have updated the original post, thanks!

**Comment by sttts** [comment]
Do the other components like the scheduler and the kubelets work behind the proxy? They should use the same mechanisms to communicate with the apiserver.

**Comment by sttts** [comment]
I see a DNS error and `403 Forbidden`. Where are those from?

**Comment by NanXiao** [comment]
@sttts 
> Do the other components like the scheduler and the kubelets work behind the proxy? They should use the same mechanisms to communicate with the apiserver.
Yes, the `apiserver`, `controller-manager` and `scheduler` all runs on the same VM (`10.100.3.41`):

	$ ps -ef | grep km
	nan       6624     1  0 Dec23 ?        00:04:41 km apiserver --address=10.100.3.41 --etcd-servers=http://10.100.3.41:4001 --service-cluster-ip-range=10.10.10.0/24 --port=8888 --cloud-provider=mesos --cloud-config=mesos-cloud.conf --secure-port=0 --v=1
	nan       6633     1  0 Dec23 ?        00:04:18 km controller-manager --master=10.100.3.41:8888 --cloud-provider=mesos --cloud-config=./mesos-cloud.conf --v=1
	nan       6646     1  0 Dec23 ?        00:01:22 km scheduler --address=10.100.3.41 --mesos-master=10.100.3.49:5050 --etcd-servers=http://10.100.3.41:4001 --mesos-user=root --api-servers=10.100.3.41:8888 --cluster-dns=10.10.10.10 --cluster-domain=cluster.local --v=2

**Comment by sttts** [comment]
`dial tcp: lookup web-proxy.corp.xx.com on 16.155.255.10:53: read udp 10.100.3.41:36156->16.155.255.10:53: i/o timeout`, port 53 is DNS usually.

**Comment by NanXiao** [comment]
@sttts : OK. Because this message occurs only once, I think it may related to our `DNS`. But following error occurs every time:  

    E1224 12:33:59.826115    7359 statusupdater.go:68] Error listing slaves without kubelet: HTTP request failed with code 403: 403 Forbidden

I try to analyze network packets, but can't find the `403 Forbidden` messages. So whether it is possible or not the `403 Forbidden` is not generated from network package?

**Comment by sttts** [comment]
I guess you have maximal logs already, i.e. log-level 9?

**Comment by NanXiao** [comment]
@sttts : Yes, I have tried log-level 8 or 10. But don't seem obvious differenence.

**Comment by jdef** [comment]
I suspect that we're seeing two connections because the controller-manager
runs the mesos cloud provider (has its own http client), which in turn uses
the ZK detector (has its own http client).

What's your mesos-cloud.conf look like?

**Comment by jdef** [comment]
was there ever any resolution for this?
